demos/cipher/aesgcm.c

```diff
@@ -122,6 +123,22 @@ static int aes_gcm_encrypt(void)
     if (!EVP_EncryptFinal_ex(ctx, outbuf, &tmplen))
         goto err;
 
+    /* Retrieve service indicator for IV generation */
+    iv_gen_rand = -1;
+    params[0] = OSSL_PARAM_construct_int(OSSL_CIPHER_PARAM_AEAD_IV_GENERATED, &iv_gen_rand);
+    if (!EVP_CIPHER_CTX_get_params(ctx, params))
+        goto err;
+    if (OSSL_PARAM_modified(params))
+        printf("Service indicator for OSSL_CIPHER_PARAM_AEAD_IV_GENERATED is: %d\n", iv_gen_rand);
```

Comment:
As this is sample code, would it be useful to check that `iv_gen_rand` and `random_iv` are either both `0` or both non-zero?

Reply:
This is end user code and shouldn't access internal context structure fields. This is documentation and example on how to use this indicator that uses public APIs only, as needed to be included or referenced in the Security Policy for this FIPS service indicator. As this indicator is set when IV length is more than 96 bits, random call was successful and stored the output.

Reply:
I'm not suggesting accessing internal context fields. `random_iv` is the argument to this function that says we want a generated (random) IV. `iv_gen_rand` is from the provider saying whether it did it or not. As the code stands it's not clear that the two should be related (i.e. if `random_iv` is set on entry to this function, then the provider should generate the IV, so `iv_gen_rand` should come back as `1`)

Reply:
@tom-cosgrove-arm it may or may not! depending on which runtime openssl the example is running against. See the sample output in this PR. As this example is universal for all openssl 3+.

That's why there is helpful text saying things.

Reply:
Would it be more clear if i make two separate encrypt functions without any arguments, such that the example is always ready to "copy & paste"?

Reply:
Would it be more clear if i make two separate encrypt functions without any arguments, such that the example is always ready to "copy & paste"?

Reply:
I know people do cut-n-paste example code, but it's not something to be encouraged! Maybe just a comment in the code with a recommendation to have the provider generate the IV? If there are two functions, people might pick either one to copy. If there's one, with comments in it, there's no choice when copying, and the comment might at least get seen in code review!